query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java

```diff
@@ -429,8 +429,10 @@ && shouldModifyIndexes(command, ctx, storedKey)) {
       }
       if (isPotentiallyIndexedType(newValue)) {
          if (shouldModifyIndexes(command, ctx, storedKey)) {
-            // This means that the entry is just modified so we need to update the indexes and not add to them.
-            operation = operation.thenCompose(r -> updateIndexes(skipIndexCleanup, newValue, key, segment));
+            // We don't need to wait for a possible removeFromIndexes operation,
+            // since if it exists, the oldValue is UNKNOWN or replacedWithADifferentEntity is true,
+            // which implies that the delete and the add operations are related to different indexes
```

Comment:
Interesting. Did you get confirmation that this was indeed the case for the person who reported the problem?

Reply:
From my test and looking at the code, in particular see the commit https://github.com/infinispan/infinispan/pull/10292/commits/3c8c832fff7b4598a33d738bf3fb0fc1f129fd11, it seems so.
I asked to the original issue reporter on the original issue Jira page.

Reply:
So given this change it makes me think that the `removeFromIndexes` operation cannot complete successfully until the `updateIndexes` invocation has occurred. Can we not just invoke `updateIndexes` first and then call `removeFromIndexes`?

Reply:
> it makes me think that the `removeFromIndexes` operation cannot complete successfully until the `updateIndexes` invocation has occurred

I'm not sure why you think that?

The two operations are unrelated since they apply to two different indexes.
They literally affect two different files on the filesystem.

They can execute in any order, so I'm not sure how what you're suggesting would change anything....

Reply:
It makes me think that because having the second index update wait on the first causes neither to complete. If they were indeed independent operations not interdependent upon each other it shouldn't matter what order you had them wait on each other at all. It just seems fishy to me.

Reply:
Before the change the only operation submitted to the indexer is `removeFromIndexes`. The other operation never does, because `removeFromIndexes` never finished. If that hangs by itself, how would adding additional operations cause it to not hang? The only way is if the operations were interdependent. That is why I am saying something must be going on between the two operations. Either that or this change wouldn't fix the issue.

Reply:
> @fax4ever , you confirm that you weren't able to reproduce the problem when using only one type? And that your patch did fix the problem exhibited in your test

Yes I do and yes it did it